### ISSUE TITLE
Option to automatically validate and provide bindable "valid" property

### ISSUE THREAD

**Issue by vsimonian** [PolymerElements/iron-input#34]
First of all, if this already is possible/exists, please, banish this issue to the shadows!

If not, though, here's what I think would be super swell:

## What

- `auto-validate` property:

```js
/**
 * If true, automatically validates the input element whenever its value is changed
 * and updates the `valid` property with the result.
 *
 * @attribute autoValidate
 * @type {boolean}
 */
autoValidate: {
  type: Boolean,
  value: false
}
```

- `valid` property:

```js
/**
 * If `autoValidate` is true, returns whether or not the input element is currently
 * in a valid state.
 *
 * @attribute valid
 * @type {boolean}
 */
valid: {
  type: Boolean,
  notify: true,
  readOnly: true
}
```

- Relevant changes in `_bindValueChanged`:

```js
_bindValueChanged: function() {
  if (this.value !== this.bindValue) {
    this.value = !this.bindValue ? '' : this.bindValue;
  }
  // manually notify because we don't want to notify until after setting value
  this.fire('bind-value-changed', {value: this.bindValue});
  if (this.autoValidate) {
    this._setValid(this.validate());
  }
},
```

## Why

Currently, validation is super easy if you're trying to validate a swath of elements. Just iterate through each one, and if anything isn't right, bark at it.

But here's a use case that currently (to my knowledge) isn't possible:

```html
<my-validator></my-validator>

<input is="iron-input" validator="my-validator" bind-value="{{value}}"
  auto-validate valid="{{valueValid}}">

<label hidden$="[[valueValid]]">
  Oh what in heavens are you thinking? That's not even <i>remotely</i> valid input.
</label>
```

Instead, I've had to resort to this:

```html
<my-validator></my-validator>

<input id="myInput" is="iron-input" validator="my-validator" bind-value="{{value}}">

<label hidden$="[[isValueValid(value)]]">
  Again with your invalid input?! Go to the corner and think about your life.
</label>

<script>
...
isValueValid: function() {
  return this.$.myInput.validate();
}
```

Yucky! :grimacing:

Though, that's not even nearly as bad as it starts to get when dealing with a bunch of elements at a time. `isValueValid` has no idea what sent it `value`. As far as it's concerned, it could be anything. So, I end up with many, many, *many* functions. Eurgh. Icky.

It would be much nicer to be able to do it all declaratively. It would greatly simplify the whole process, thus making it faster to implement, and (in my opinion) more user-friendly.

**Comment by notwaldorf** [comment]
Have you checked out `paper-input`? That already has an `auto-validate` property. You can use it like this:
```
<paper-input label="only type letters (auto-validate)" auto-validate pattern="[a-zA-Z]*" error-message="letters only!"></paper-input>
```

Is there a reason why you want it directly on `iron-input`? Is it because you don't want the paper style that comes with `paper-input`?

**Comment by vsimonian** [comment]
`paper-input` is fantastic! Sometimes, though, it feels like bringing everything and the kitchen sink.

I think that having `validate()` begs the need for a feature such as `auto-validate,` which is far from being a complex addition that would turn `iron-input` into something other than the simple element that it is. Many times when I (and I would like to assume others) would prefer to use an `iron-input` instead of a `paper-input`, it'd be because

- `iron-input` has no UI preference
- they're throwing together a very lean/lightweight app/page and trying not to bring along more code than is necessary

And in those cases, instead of having to wire up some odd handler to get validation up and running, or dragging `paper-input` in with all its styles and dependencies, it'd be much simpler to throw in an `iron-input`, set `auto-validate` to true, and done! If the user wasn't using `hidden$=` but a CSS rule targeting the `iron-input`'s `[invalid]` instead, they wouldn't even need a `dom-bind` template!

So, that's my two cents. I think `auto-validate` could be a super nifty thing to have on `iron-input`, and it wouldn't be bringing too much baggage along. I'm certainly not asking that `paper-input-behavior` gets shifted down to `iron-input`, that'd be too much. But, `auto-validate`? I think that the cost of throwing it in far outweighs the cost of not having it.

But that's just me, and I don't make the decisions 'round here (until I start bribing everyone with free coffee).

**Comment by notwaldorf** [comment]
If this isn't urgent, I'd like to wait a bit and talk it over with @morethanreal before we decide on anything.

**Comment by vsimonian** [comment]
It's certainly not urgent! Better to make a well-thought-out decision instead of rushing into anything, so I'm all for more discussion.

**Comment by kshep92** [comment]
I agree with this one. I don't always want to use the paper elements, so a 'validate-on-change' feature would really be nice, especially since I'm a big fan of how easy it is to hide/show error messages in AngularJS.
